File: Features/confidentialidentity-whistleblower/contracts/src/test/java/net/corda/samples/whistleblower/contracts/BlowWhistleContractTests.java
package net.corda.samples.whistleblower.contracts;

import net.corda.core.contracts.Contract;
import net.corda.core.identity.CordaX500Name;
import net.corda.samples.whistleblower.states.BlowWhistleState;
import net.corda.testing.core.DummyCommandData;
import net.corda.testing.core.TestIdentity;
import net.corda.testing.node.MockServices;
import org.junit.Test;

import java.util.Arrays;

import static net.corda.testing.node.NodeTestUtils.transaction;

public class BlowWhistleContractTests {

    private final TestIdentity a = new TestIdentity(new CordaX500Name("alice", "", "GB"));
    private final TestIdentity b = new TestIdentity(new CordaX500Name("bob", "", "GB"));
    private final TestIdentity c = new TestIdentity(new CordaX500Name("bad corp", "", "GB"));

    private MockServices ledgerServices = new MockServices(
            Arrays.asList("net.corda.samples.whistleblower.contracts")
    );

    BlowWhistleState st = new BlowWhistleState(c.getParty(), a.getParty().anonymise(), b.getParty().anonymise());

    @Test
    public void SantaSessionContractImplementsContract() {
        assert (new BlowWhistleContract() instanceof Contract);
    }

    @Test
    public void contractRequiresZeroInputsInTheTransaction() {

        transaction(ledgerServices, tx -> {
            // Has an input, will fail.
            tx.output(BlowWhistleContract.ID, st);
            tx.output(BlowWhistleContract.ID, st);
            tx.command(Arrays.asList(a.getPublicKey(), b.getPublicKey()), new BlowWhistleContract.Commands.BlowWhistleCmd());
            tx.fails();
            return null;
        });

        transaction(ledgerServices, tx -> {
            // Has no input, will verify.
            tx.output(BlowWhistleContract.ID, st);
            tx.command(Arrays.asList(a.getPublicKey(), b.getPublicKey()), new BlowWhistleContract.Commands.BlowWhistleCmd());
            tx.verifies();
            return null;
        });
    }

    @Test
    public void contractRequiresSpecificCommand() {
        transaction(ledgerServices, tx -> {
            // Has wrong command type, will fail.
            tx.output(BlowWhistleContract.ID, st);
            tx.command(Arrays.asList(a.getPublicKey()), DummyCommandData.INSTANCE);
            tx.fails();
            return null;
        });

        transaction(ledgerServices, tx -> {
            // Has correct command type, will verify.
            tx.output(BlowWhistleContract.ID, st);
            tx.command(Arrays.asList(a.getPublicKey(), b.getPublicKey()), new BlowWhistleContract.Commands.BlowWhistleCmd());
            tx.verifies();
            return null;
        });

    }

}

File: kotlin-samples/Features/confidentialIdentity-whistleblower/contracts/src/test/kotlin/net/corda/samples/whistleblower/contracts/BlowWhistleContractTests.kt
package net.corda.samples.whistleblower.contracts

import net.corda.core.contracts.CommandData
import net.corda.core.contracts.Contract
import net.corda.core.contracts.TypeOnlyCommandData
import net.corda.core.identity.CordaX500Name
import net.corda.samples.whistleblower.states.BlowWhistleState
import net.corda.testing.core.TestIdentity
import net.corda.testing.node.MockServices
import net.corda.testing.node.ledger
import org.junit.Test
import java.util.*


class BlowWhistleContractTests {
    private val a = TestIdentity(CordaX500Name("alice", "", "GB"))
    private val b = TestIdentity(CordaX500Name("bob", "", "GB"))
    private val c = TestIdentity(CordaX500Name("bad corp", "", "GB"))

    private val ledgerServices = MockServices(
            Arrays.asList("net.corda.samples.whistleblower.contracts")
    )

    var st: BlowWhistleState = BlowWhistleState(c.party, a.party.anonymise(), b.party.anonymise())

    // A pre-defined dummy command.
    interface Commands : CommandData {
        class DummyCommand : TypeOnlyCommandData(), Commands
    }


    @Test
    fun SantaSessionContractImplementsContract() {
        assert(BlowWhistleContract() is Contract)
    }

    @Test
    fun contractRequiresZeroInputsInTheTransaction() {
        ledgerServices.ledger {
            transaction {
                // Has an input, will fail.
                output(BLOW_WHISTLE_CONTRACT_ID, st)
                output(BLOW_WHISTLE_CONTRACT_ID, st)
                command(listOf(a.publicKey, b.publicKey), BlowWhistleContract.Commands.BlowWhistleCmd())
                fails()
            }

            transaction {
                // Has no input, will verify.
                output(BLOW_WHISTLE_CONTRACT_ID, st)
                command(listOf(a.publicKey, b.publicKey), BlowWhistleContract.Commands.BlowWhistleCmd())
                verifies()
            }
        }
    }

    @Test
    fun contractRequiresSpecificCommand() {
        ledgerServices.ledger {
            transaction {
                // Has wrong command type, will fail.
                output(BLOW_WHISTLE_CONTRACT_ID, st)
                command(Arrays.asList(a.publicKey, b.publicKey), Commands.DummyCommand())
                fails()
            }

            transaction {
                // Has correct command type, will verify.
                output(BLOW_WHISTLE_CONTRACT_ID, st)
                command(Arrays.asList(a.publicKey, b.publicKey), BlowWhistleContract.Commands.BlowWhistleCmd())
                verifies()
            }
        }
    }
}
